ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/MessageListenerProxy.java

```diff
@@ -0,0 +1,74 @@
+package org.jboss.narayana.jta.jms;
+
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.jms.Message;
+import javax.jms.MessageListener;
+import javax.jms.XASession;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.Transactional;
+import java.lang.reflect.Method;
+
+public class MessageListenerProxy implements MessageListener {
+
+    private MessageListener messageListener;
+    private TransactionManager transactionManager;
+    private XASession xaSession;
+
+    static boolean isDeclaredTransactional(MessageListener messageListener) {
+        try {
+            if (messageListener == null) {
+                return false;
+            }
+
+            Class<? extends MessageListener> listenerClass = messageListener.getClass();
+            if (listenerClass.isAnnotationPresent(Transactional.class)) {
+                return true;
+            } else {
+                Method onMessageMethod = listenerClass.getDeclaredMethod("onMessage", Message.class);
+                return onMessageMethod.isAnnotationPresent(Transactional.class);
+            }
+        } catch (NoSuchMethodException e) {
+            // Should not occur, as messageListener is an instance of javax.jms.MessageListener
+        }
+        return false;
+    }
+
+    public MessageListenerProxy(MessageListener messageListener, XASession session) {
+        this(messageListener, session, com.arjuna.ats.jta.TransactionManager.transactionManager());
+    }
+
+    /**
+     * Mainly for tests
+     */
+    MessageListenerProxy(MessageListener messageListener, XASession session, TransactionManager transactionManager ) {
+        this.messageListener = messageListener;
+        this.transactionManager = transactionManager;
+        this.xaSession = session;
+    }
+
+    @Override
+    public void onMessage(Message jmsMessage) {
```

Comment:
FYI this is a Martyn's reply on the mailing list http://activemq-users.markmail.org/thread/zqm34i32nl3uqmqb#query:+page:1+mid:ez67zqkj7lbn2lgj+state:results.

I had a chat with him yesterday on IRC too. He said that most likely this scenario will not work with other vendors since it's not in the spec. The reason it works now is because of the way Artemis is implemented.